source/fuzz/added_function_reducer.cpp

```diff
@@ -0,0 +1,284 @@
+  // module.
+  auto ir_context =
+      BuildModule(target_env_, consumer_, binary_under_reduction.data(),
+                  binary_under_reduction.size());
+  assert(ir_context != nullptr && "The binary should be parsable.");
+  for (auto& type_or_value : ir_context->module()->types_values()) {
+    if (type_or_value.opcode() != SpvOpVariable) {
+      continue;
+    }
+    if (irrelevant_pointee_global_variables.count(type_or_value.result_id())) {
+      continue;
+    }
+    if (!ir_context->get_def_use_mgr()->WhileEachUse(
+            &type_or_value,
+            [this, &ir_context](opt::Instruction* user,
+                                uint32_t /*unused*/) -> bool {
+              auto block = ir_context->get_instr_block(user);
+              if (block != nullptr &&
+                  block->GetParent()->result_id() == GetAddedFunctionId()) {
+                return false;
+              }
+              return true;
+            })) {
+      return false;
+    }
+  }
+
+  // It must be possible to successfully apply all the transformations, with
+  // the transformation at index |index_of_add_function_transformation_|
+  // simplified to use the version of the added function from
+  // |binary_under_reduction|.
+  std::vector<uint32_t> binary_out;
+  protobufs::TransformationSequence modified_transformations;
+  ReplayAdaptedTransformations(binary_under_reduction, &binary_out,
+                               &modified_transformations);
+  if (transformation_sequence_in_.transformation_size() !=
+      modified_transformations.transformation_size()) {
+    return false;
+  }
+
+  // The resulting binary must be deemed interesting according to the shrinker's
+  // interestingness function.
+  return shrinker_interestingness_function_(
+      binary_out, counter_for_shrinker_interestingness_function);
+}
+
+void AddedFunctionReducer::ReplayPrefixAndAddFunction(
+    std::vector<uint32_t>* binary_out,
+    std::unordered_set<uint32_t>* irrelevant_pointee_global_variables) const {
+  assert(transformation_sequence_in_
+             .transformation(index_of_add_function_transformation_)
+             .has_add_function() &&
+         "A TransformationAddFunction is required at the given index.");
+
+  auto replay_result = Replayer(target_env_, consumer_, binary_in_,
+                                initial_facts_, transformation_sequence_in_,
+                                index_of_add_function_transformation_,
+                                validate_during_replay_, validator_options_)
+                           .Run();
+  assert(replay_result.status == Replayer::ReplayerResultStatus::kComplete &&
+         "Replay should succeed");
+  assert(static_cast<uint32_t>(
+             replay_result.applied_transformations.transformation_size()) ==
+             index_of_add_function_transformation_ &&
+         "All requested transformations should have applied.");
+
+  auto ir_context = std::move(replay_result.transformed_module);
```

Comment:
Nit: can we just get a reference to the module here, instead of moving?

Reply:
Done.